client/landing/custom-editor/utils.ts

```diff
@@ -0,0 +1,110 @@
+/**
+ * External dependencies
+ */
+import apiFetch from '@wordpress/api-fetch';
+import wpcomRequest from 'wpcom-proxy-request';
+
+/**
+ * A lot of the following is copied from D37093-code.
+ * Maybe we can move this code into Calypso `packages/`, parametrize it (e.g. apiFetch)
+ * and move into Calypso's `packages/wpcom-proxy-request`, or publish as a new npm for use on WP.com.
+ */
+
+// FIXME -- quick'n'dirty setting of a few params
+const _currentSiteId = 89417913; // @ockham's test site, replace with your own
+
+const wpApiSettings = {
+	root: 'https://public-api.wordpress.com/',
+};
+
+/**
+ * Normalizes the path for requests to the public API.
+ */
+function wpcomFetchNormalizePath( options ) {
+	if ( options.url && options.url.indexOf( wpApiSettings.root ) !== -1 ) {
+		options.path = options.url.replace( wpApiSettings.root, '' );
+		delete options.url;
+	}
+
+	if ( options.path ) {
+		// Ensures path starts with a slash.
+		if ( ! options.path.startsWith( '/' ) ) {
+			options.path = '/' + options.path;
+		}
+
+		// Removes namespace from path.
+		if ( options.apiNamespace ) {
+			options.path = options.path.replace( '/' + options.apiNamespace, '' );
+		}
+	}
+
+	return options;
+}
+
+/**
+ * Creates a fetch-like response.
+ */
+function wpcomFetchCreateFetchResponse( data, status, headers ) {
+	var fetchResponse;
+	var normalizedHeaders = {};
+	for ( var header in headers ) {
+		normalizedHeaders[ header.toLowerCase() ] = headers[ header ];
+	}
+	if ( Array.isArray( data ) ) {
+		fetchResponse = data.slice( 0 );
+	} else {
+		fetchResponse = Object.assign( {}, data );
+	}
+	fetchResponse.status = status;
+	fetchResponse.json = function() {
+		return Promise.resolve( data );
+	};
+	fetchResponse.headers = {
+		get: function( name ) {
+			return normalizedHeaders[ name && name.toLowerCase() ];
+		},
+	};
+	return fetchResponse;
+}
+
+/**
+ * Determines the namespace from the request path.
+ */
+function wpcomFetchSetNamespace( options, next ) {
+	if ( options.path && ! options.namespace ) {
+		var namespace = options.path.match( /^\/([a-z]+\/v?[0-9.]+)\// );
+		if ( namespace ) {
+			options.apiNamespace = namespace[ 1 ];
+			options.path = options.path.replace( '/' + options.apiNamespace, '' );
+		} else {
+			// Defaults all requests to the wp/v2 namespace.
+			options.apiNamespace = 'wp/v2';
+		}
+	}
+	console.log( 'post-norm options', options );
+	return next( options, next );
+}
+
+/**
+ * Prefixes any non-global request endpoint to the site specific endpoint.
+ */
+function wpcomFetchAddSitePrefix( options, next ) {
+	if ( options.path && ! options.global && options.path.indexOf( '/sites/' ) === -1 ) {
+		options.path = '/sites/' + _currentSiteId + options.path;
```

Comment:
This whole module is very interesting. Any ideas how we would be able to get this to work server-side? Where would the middleware be able to get the current site ID from? Is this already solved elsewhere?

Reply:
> This whole module is very interesting. Any ideas how we would be able to get this to work server-side?

Could you elaborate? We could probably implement those endpoints in PHP and pass through to the 'actual' ones. I'm not quite sure why we would do that, since it seems it would require us to keep a much more explicit list of existing endpoints (which would mean higher maintenance burden etc) :thinking: 

> Where would the middleware be able to get the current site ID from? Is this already solved elsewhere?

I lifted this stuff mostly from the `gutenberg-wpcom` plugin, and that's using a `_currentSiteId` global that's attached to `window` somehow by another plugin (I think).

Alternatively, I think that on the client side, we should be able to store the current site ID somewhere in Redux/`@wordpress/data` state and have the middleware pull it from there.

Reply:
By "server side" I meant the Calypso Node server, in the case of SSR, for example.